vfssimple/vfssimple.go

```diff
@@ -31,78 +37,81 @@ func NewLocation(uri string) (vfs.Location, error) {
 func NewFile(uri string) (vfs.File, error) {
 	fs, host, path, err := parseSupportedURI(uri)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("unable to create vfs.File for uri %q: %w", uri, err)
 	}
 
 	return fs.NewFile(host, path)
 }
 
-func parseSupportedURI(uri string) (vfs.FileSystem, string, string, error) {
-	var err error
+// parseURI attempts to parse a URI and validate that it returns required results
+func parseURI(uri string) (scheme, authority, path string, err error) {
+	// return early if blank uri
+	if uri == "" {
+		err = ErrBlankURI
+		return
+	}
+
+	// parse URI
 	var u *url.URL
 	u, err = url.Parse(uri)
 	if err != nil {
-		return nil, "", "", err
-	}
-	host := u.Host
-	path := u.Path
-	user := u.User
-	if user.String() != "" {
-		host = fmt.Sprintf("%s@%s", user, host)
+		err = fmt.Errorf("unknown url.Parse error: %w", err)
+		return
 	}
 
-	var fs vfs.FileSystem
-	for _, backendScheme := range backend.RegisteredBackends() {
-		// Azure
-		if azure.IsValidURI(u) {
-			host, path, err = azure.ParsePath(u.Path)
-		}
-
-		// Object-level backend
-		if strings.HasPrefix(uri, backendScheme) {
-			fs = backend.Backend(backendScheme)
-			break
-		}
-		// Bucket-level backend
-		volume := fmt.Sprintf("%s://%s/", u.Scheme, u.Host)
-		if volume == backendScheme {
-			fs = backend.Backend(backendScheme)
-			break
-		}
-
-		// Location-level backend
-		if isInPath(u, volume, backendScheme) {
-			fs = backend.Backend(backendScheme)
-			break
-		}
+	// validate schema
+	scheme = u.Scheme
+	if u.Scheme == "" {
+		err = ErrMissingScheme
+		return
+	}
 
-		// Scheme-level backend
-		if u.Scheme == backendScheme {
-			fs = backend.Backend(backendScheme)
-		}
+	// validate authority
+	authority = u.Host
+	path = u.Path
+	if azure.IsValidURI(u) {
+		authority, path, err = azure.ParsePath(path)
 	}
 
-	if fs == nil {
-		err = fmt.Errorf("%s is an unsupported uri scheme", u.Scheme)
+	if u.User.String() != "" {
+		authority = fmt.Sprintf("%s@%s", u.User, u.Host)
+	}
+	// network-based schemes require authority, but not file:// or mem://
+	if authority == "" && !(scheme == "file" || scheme == "mem") {
```

Comment:
These should probably reference the constants in each of the supplied backends:

`backend.os.Scheme` and `backends.mem.Scheme`